### PR TITLE
Fix sqlglot>=23.0.0 issue

### LINKED ISSUE #2072: Sqlglot 23.0.0 breaks EM Training
### What happens?

Via `_get_comparison_levels_corresponding_to_training_blocking_rule` and `ComparisonLevel._is_exact_match` we end up with the error.

```sh
if identifiers[0] == identifiers[1]:
       ~~~~~~~~~~~^^^
IndexError: list index out of range
```

Working on v22.2.1 (as in lockfile) - not sure of the exact `sqlglot` version where this breaks

### To Reproduce

With `sqlglot==23.0.0`:

```py
import splink.duckdb.comparison_library as cl
from splink.duckdb.linker import DuckDBLinker
from splink.datasets import splink_datasets

settings = {
    "link_type": "dedupe_only",
    "unique_id_column_name": "unique_id",
    "comparisons": [
        cl.levenshtein_at_thresholds(
            "first_name", [2], term_frequency_adjustments=True
        ),
        cl.levenshtein_at_thresholds("surname", [1], term_frequency_adjustments=True),
    ],
}

df = splink_datasets.fake_1000

linker = DuckDBLinker(df, settings)
blocking_rule = "l.first_name = r.first_name"
training_session_fn = linker.estimate_parameters_using_expectation_maximisation(
    blocking_rule
)
```

### OS:

Debian

### Splink version:

3.9.13

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

### LINKED ISSUE #2076: ERROR - IndexError: list index out of range
### Discussed in https://github.com/moj-analytical-services/splink/discussions/2075

<div type='discussions-op-text'>

<sup>Originally posted by **meharc** March 19, 2024</sup>
I replicated an example code, identical to the one provided in this repository, within a Databricks notebook to explore Splink.

```
from splink.duckdb.linker import DuckDBLinker
import splink.duckdb.comparison_library as cl
import splink.duckdb.comparison_template_library as ctl
from splink.duckdb.blocking_rule_library import block_on
from splink.datasets import splink_datasets

df = splink_datasets.fake_1000

settings = {
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        block_on("first_name"),
        block_on("surname"),
    ],
    "comparisons": [
        ctl.name_comparison("first_name"),
        ctl.name_comparison("surname"),
        ctl.date_comparison("dob", cast_strings_to_date=True),
        cl.exact_match("city", term_frequency_adjustments=True),
        ctl.email_comparison("email", include_username_fuzzy_level=False),
    ],
}

linker = DuckDBLinker(df, settings)
linker.estimate_u_using_random_sampling(max_pairs=1e6)

blocking_rule_for_training = block_on(["first_name", "surname"])

linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training, estimate_without_term_frequencies=True)

blocking_rule_for_training = block_on("dob")
linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training, estimate_without_term_frequencies=True)

pairwise_predictions = linker.predict()

clusters = linker.cluster_pairwise_predictions_at_threshold(pairwise_predictions, 0.95)
`clusters.as_pandas_dataframe(limit=5)```
```

This gives me the following error:

```
IndexError: list index out of range
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
File <command-1077583777744438>, line 29
     25 linker.estimate_u_using_random_sampling(max_pairs=1e6)
     27 blocking_rule_for_training = block_on(["first_name", "surname"])
---> 29 linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training, estimate_without_term_frequencies=True)
     31 blocking_rule_for_training = block_on("dob")
     32 linker.estimate_parameters_using_expectation_maximisation(blocking_rule_for_training, estimate_without_term_frequencies=True)

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-56a9ba29-5210-4bf7-abfd-49580965d839/lib/python3.10/site-packages/splink/linker.py:1671, in Linker.estimate_parameters_using_expectation_maximisation(self, blocking_rule, comparisons_to_deactivate, comparison_levels_to_reverse_blocking_rule, estimate_without_term_frequencies, fix_probability_two_random_records_match, fix_m_probabilities, fix_u_probabilities, populate_probability_two_random_records_match_from_trained_values)
   1659     if comparison_levels_to_reverse_blocking_rule is None:
   1660         logger.warning(
   1661             "\nWARNING: \n"
   1662             "You have provided comparisons_to_deactivate but not "
   (...)
   1668             "as an exact match."
   1669         )
-> 1671 em_training_session = EMTrainingSession(
   1672     self,
   1673     blocking_rule,
   1674     fix_u_probabilities=fix_u_probabilities,
   1675     fix_m_probabilities=fix_m_probabilities,
   1676     fix_probability_two_random_records_match=fix_probability_two_random_records_match,  # noqa 501
   1677     comparisons_to_deactivate=comparisons_to_deactivate,
   1678     comparison_levels_to_reverse_blocking_rule=comparison_levels_to_reverse_blocking_rule,  # noqa 501
   1679     estimate_without_term_frequencies=estimate_without_term_frequencies,
   1680 )
   1682 em_training_session._train()
   1684 self._populate_m_u_from_trained_values()

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-56a9ba29-5210-4bf7-abfd-49580965d839/lib/python3.10/site-packages/splink/em_training_session.py:70, in EMTrainingSession.__init__(self, linker, blocking_rule_for_training, fix_u_probabilities, fix_m_probabilities, fix_probability_two_random_records_match, comparisons_to_deactivate, comparison_levels_to_reverse_blocking_rule, estimate_without_term_frequencies)
     66     self._comparison_levels_to_reverse_blocking_rule = (
     67         comparison_levels_to_reverse_blocking_rule
     68     )
     69 else:
---> 70     self._comparison_levels_to_reverse_blocking_rule = self._original_settings_obj._get_comparison_levels_corresponding_to_training_blocking_rule(  # noqa
     71         blocking_rule_for_training.blocking_rule_sql
     72     )
     74 self._settings_obj._probability_two_random_records_match = (
     75     self._blocking_adjusted_probability_two_random_records_match
     76 )
     78 self._training_fix_u_probabilities = fix_u_probabilities

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-56a9ba29-5210-4bf7-abfd-49580965d839/lib/python3.10/site-packages/splink/settings.py:338, in Settings._get_comparison_levels_corresponding_to_training_blocking_rule(self, blocking_rule)
    336 for cc in ccs:
    337     for cl in cc.comparison_levels:
--> 338         if cl._is_exact_match:
    339             exact_comparison_levels.append(cl)
    341 # Where exact match on multiple columns exists, use that instead of individual
    342 # exact match columns
    343 # So for example, if we have a param estimate for exact match on first name AND
    344 # surname, prefer that
    345 # over individual estimtes for exact match first name and surname.

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-56a9ba29-5210-4bf7-abfd-49580965d839/lib/python3.10/site-packages/splink/comparison_level.py:503, in ComparisonLevel._is_exact_match(self)
    501 exprs = _get_and_subclauses(sql_cnf)
    502 for expr in exprs:
--> 503     if not _is_exact_match(expr):
    504         return False
    505 return True

File /local_disk0/.ephemeral_nfs/envs/pythonEnv-56a9ba29-5210-4bf7-abfd-49580965d839/lib/python3.10/site-packages/splink/comparison_level.py:45, in _is_exact_match(sql_syntax_tree)
     43     if type(subtree) is Identifier:
     44         identifiers.append(subtree.this[:-2])
---> 45 if identifiers[0] == identifiers[1]:
     46     return True
     47 else:

IndexError: list index out of range
```

I've just copied and pasted the code from the tutorial located on the homepage of the repository. Do you have any suggestions or insights into why this might be occurring?</div>

### PR DESCRIPTION
**Pull request by RobinL**
Closes #2072, see that issue for details 
Closes #2075 
Closes #2076

### CODE REVIEW

**Comment by ADBond** [comment]
I think there will also be similar updates needed to [SQLTask](https://github.com/moj-analytical-services/splink/blob/bfaf8d142f7aad03f281b6324dd530e202eaec9a/splink/pipeline.py#L26) and [_exact_match_colname](https://github.com/moj-analytical-services/splink/blob/bfaf8d142f7aad03f281b6324dd530e202eaec9a/splink/comparison_level.py#L59)

**Comment by RobinL** [comment]
Oops - that's me being overly optimistic, and also forgetting the tests are running against old SQL glot so of course they pass. Good spot.

I've updated the other methods, and updated the lockfile.

You can see in CI that tests passed for both the old and the new lockfile so i think we should be covered.

**Review by ThomasHepworth — APPROVED** [review]
Thanks Robin, this looks great.

Just to confirm, I've tested the code against the following SQLglot versions:
* 23.0.0
* 19.0.0
* 13.0.0

Everything is working as expected.

File: splink/comparison_level.py [hunk 1]
```diff
@@ -37,12 +37,9 @@ def _is_exact_match(sql_syntax_tree):
...
 
-    identifiers = []
-    for tup in sql_syntax_tree.walk():
-        subtree = tup[0]
-        if type(subtree) is Identifier:
-            identifiers.append(subtree.this[:-2])
-    if identifiers[0] == identifiers[1]:
+    identifiers = list(sql_syntax_tree.find_all(Identifier))
+    idents_truncated = [i.this[:-2] for i in identifiers]
+    if idents_truncated[0] == idents_truncated[1]:
```
[the remark this answers is not in the document]
Comment (RobinL) [reply splink/comparison_level.py]: Good idea!

**Pull request merged**